File: src/RSA.java
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

/**
 * Created by jonat on 21/12/2016.
 */
public class RSA {
    public static int KEY_SIZE = 512;

    public static BigInteger encrypt(BigInteger x, BigInteger e, BigInteger n) {
        return x.modPow(e, n);
    }

    public static BigInteger encrypt(BigInteger x, ArrayList<BigInteger> pub) {
        return x.modPow(pub.get(1), pub.get(0));
    }

    public static BigInteger decrypt(BigInteger x, BigInteger d, BigInteger n) {
        return x.modPow(d, n);
    }

    public static BigInteger decrypt(BigInteger x, ArrayList<BigInteger> priv) {
        return x.modPow(priv.get(1), priv.get(0));
    }

    public static boolean primalityTestFermat(BigInteger m) {
        BigInteger nn = new BigInteger("2");
        return nn.modPow(m.subtract(new BigInteger("1")), m).equals(new BigInteger("1"));
    }

    public static HashMap<String, BigInteger> keygen() {
        BigInteger e = new BigInteger(16, 2, new Random());
        while (!primalityTestFermat(e)) {
            e = e.nextProbablePrime();
        }

        HashMap<String, BigInteger> result = new HashMap<>();
        try {
            result = keygen(e);
        } catch (Exception ex)
        {
            System.err.println("Fatal error in keygen, e was not primal");
        }

        return result;
    }

    public static HashMap<String, BigInteger> keygen(BigInteger e) throws Exception{
        if(!primalityTestFermat(e))
            throw new Exception("e must be primal!");

        HashMap<String, BigInteger> keygen = new HashMap<>();

        BigInteger phi_n, n, d, p, q;

        do {
            p = new BigInteger(KEY_SIZE, 2, new Random());
            while (!primalityTestFermat(p)) {
                p = p.nextProbablePrime();
            }
            q = new BigInteger(KEY_SIZE, 2, new Random());
            while (!primalityTestFermat(q)) {
                q = q.nextProbablePrime();
            }
            n = p.multiply(q);
            phi_n = p.subtract(new BigInteger("1")).multiply(q.subtract(new BigInteger("1")));

        } while (!phi_n.gcd(e).equals(new BigInteger("1")));

        d = e.modInverse(phi_n);

        keygen.put("n", n);
        keygen.put("e", e);
        keygen.put("d", d);

        return keygen;
    }

}

File: src/main/java/RSA.kt
package main.java

import java.math.BigInteger
import java.util.*

object RSA {
    val KEY_SIZE = 512

    val ONE = BigInteger("1")
    val TWO = BigInteger("2")

    val encoder = Base64.getEncoder()

    class PublicKey constructor(val e: BigInteger, val n: BigInteger){
        fun toBase64() : String {
            return encoder.encodeToString(e.toByteArray() + n.toByteArray())
        }
    }
    class PrivateKey constructor(val d: BigInteger, val n: BigInteger)

    class NotPrimalNumberException : Exception("Number given is not a primal number")

    fun encrypt(x: BigInteger, e: BigInteger, n: BigInteger): BigInteger {
        return x.modPow(e, n)
    }

    fun encrypt(x: BigInteger, pub: PublicKey): BigInteger {
        println("$x is being encrypted with ${pub.toBase64()}")
        return encrypt(x, pub.e, pub.n)
    }

    fun decrypt(x: BigInteger, d: BigInteger, n: BigInteger): BigInteger {
        return x.modPow(d, n)
    }

    fun decrypt(x: BigInteger, priv: PrivateKey): BigInteger {
        return decrypt(x, priv.d, priv.n)
    }

    fun primalTestFermat(m: BigInteger): Boolean {
        return TWO.modPow(m - ONE, m) == ONE
    }

    fun keygen(): Pair<PublicKey, PrivateKey> {
        var e = BigInteger(16, 2, Random())

        while (!primalTestFermat(e)) {
            e = e.nextProbablePrime()
        }

        return keygen(e)
    }

    @Throws(NotPrimalNumberException::class)
    fun keygen(e: BigInteger): Pair<PublicKey, PrivateKey> {
        if (!primalTestFermat(e))
            throw NotPrimalNumberException()

        var phi_n: BigInteger
        var n: BigInteger
        var p: BigInteger
        var q: BigInteger

        do {
            p = BigInteger(KEY_SIZE, 2, Random())
            while (!primalTestFermat(p)) p = p.nextProbablePrime()
            q = BigInteger(KEY_SIZE, 2, Random())
            while (!primalTestFermat(q)) q = q.nextProbablePrime()
            n = p * q
            phi_n = (p - ONE) * (q - ONE)

        } while (phi_n.gcd(e) != ONE)

        val d = e.modInverse(phi_n)

        val pub = PublicKey(e,n)
        println("A pair of keys was generated\nPublic key: ${pub.toBase64()}")

        return Pair(pub, PrivateKey(d,n))
    }

}
